File: src/app/types/timed-event.lua
local ____exports = {}
____exports.TimedEvent = {}
local TimedEvent = ____exports.TimedEvent
TimedEvent.name = "TimedEvent"
TimedEvent.__index = TimedEvent
TimedEvent.prototype = {}
TimedEvent.prototype.__index = TimedEvent.prototype
TimedEvent.prototype.constructor = TimedEvent
function TimedEvent.new(...)
    local self = setmetatable({}, TimedEvent.prototype)
    self:____constructor(...)
    return self
end
function TimedEvent.prototype.____constructor(self, func, time, safe)
    if safe == nil then
        safe = true
    end
    self.safe = safe
    self.time = time
    self.func = func
end
function TimedEvent.prototype.setTickRate(self, msPerTick)
    self.time = self.time / msPerTick
end
function TimedEvent.prototype.AttemptAction(self, currentTick, teq)
    if not self.endtime then
        self.endtime = ((currentTick + self.time) % 100000) - 1
        if self.endtime < 0 then
            self.endtime = 0
        end
    end
    if self.endtime <= currentTick then
        self:func()
        return true
    end
    return false
end
return ____exports

File: src/app/types/timed-event.ts
import { TimedEventQueue } from './timed-event-queue';

export class TimedEvent {
    private time: number;
    private func: () => boolean;
    private safe: boolean;
    private endtime: number | undefined;
    // private created: number | undefined;

    /**
     * 
     * @param func 
     * @param time 
     * @param safe wait time in ms
     */
    constructor(func: () => boolean, time: number, safe: boolean = true) {
        this.safe = safe;
        this.time = time;
        this.func = () => func();
    }

    public setTickRate(msPerTick: number) {
        this.time = this.time / msPerTick;
    }

    public AttemptAction(currentTick: number, teq: TimedEventQueue): boolean {
        if (!this.endtime) {
            this.endtime = ((currentTick + this.time) % 100000) - 1;
            if (this.endtime < 0) {
                this.endtime = 0;
            }
        }
        if (this.endtime <= currentTick) {
            // if (this.safe) {
            //     teq.GetGame().safeEventQueue.AddMed(this.func);
            // } else {
            //     teq.GetGame().eventQueue.AddMed(this.func);
            // }
            this.func();
            return true;
        }
        return false;
    }

    public GetEndTick() {
        return this.endtime || 1;
    }

    // public SetCreatedTime(time: number) {
    //     this.created = time;
    // }

    // public GetCreatedTime() {
    //     return this.created;
    // }
}